src/Compilers/CSharp/Test/Semantic/Semantics/NameOfTests.cs

```diff
@@ -1508,5 +1508,168 @@ public void nameof(string x)
+        public void TestCannotReferenceInstanceMembersFromAttributeInNameofInCSharp9()
+        {
+            var source = @"
+public class C
+{
+    [System.Obsolete(nameof(S.Length))]
+    public int P { get; }
+    public string S { get; }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics(
+                // (4,29): error CS8652: The feature 'nameof can access instance members in all contexts' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     [System.Obsolete(nameof(S.Length))]
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "S").WithArguments("nameof can access instance members in all contexts").WithLocation(4, 29));
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCannotReferenceInstanceMembersFromConstructorInitializersInNameofInCSharp9()
+        {
+            var source = @"
+public class C
+{
+    public C(string s){}
+    public C() : this(nameof(S.Length)){}
+    public string S { get; }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics(
+                // (5,30): error CS8652: The feature 'nameof can access instance members in all contexts' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public C() : this(nameof(S.Length)){}
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "S").WithArguments("nameof can access instance members in all contexts").WithLocation(5, 30));
+        }
+
+        [Fact, WorkItem(40229, "https://github.com/dotnet/roslyn/issues/40229")]
+        public void TestCannotAccessStructInstancePropertyInLambdaInNameofInCSharp9()
+        {
+            var source = @"
+using System;
+
+public struct S
+{
+    public string P { get; }
+    public void M(ref string x)
+    {
+        Func<string> func = () => nameof(P.Length);
+    }
+}";
+            CreateCompilation(source, parseOptions: TestOptions.Regular9).VerifyDiagnostics(
+                // (9,42): error CS8652: The feature 'nameof can access instance members in all contexts' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         Func<string> func = () => nameof(P.Length);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "P").WithArguments("nameof can access instance members in all contexts").WithLocation(9, 42));
+        }
+
+        [Fact]
+        public void TestCanAccessRefParameterInLambdaInNameof()
+        {
+            var source = @"
+using System;
+
+public struct S
+{
+    public void M(ref string x)
+    {
+        Func<string> func = () => nameof(x.Length);
+    }
+}";
+            CreateCompilation(source).VerifyDiagnostics();
+        }
```

Comment:
Need some tests for other types of members, such as methods, fields, events, etc.